MagicBytesValidator/Models/FileByteFilter.cs

```diff
@@ -0,0 +1,146 @@
+        
+        MimeTypes = mimeTypes;
+        Extensions = extensions;
+    }
+    
+    public class ByteCheck(int offset, byte?[] bytesToCheck)
+    {
+        public int Offset = offset;
+        public readonly byte?[] ByteArray = bytesToCheck;
+    }
+    
+    public bool Matches(byte[] fileByteStream)
+    {
+        foreach (var neededByteCheck in _neededByteChecks)
+        {
+            if (!CheckBytes(neededByteCheck, fileByteStream))
+                return false;
+        }
+        
+        foreach (var oneOf in _oneOfEachByteChecks)
+        {
+            if (!oneOf.Any(byteToCheck => CheckBytes(byteToCheck, fileByteStream)))
+            {
+                return false;
+            }
+        }
+
+        // Then check byteArrays without fixed offsets
+        // mainly byteArrays from Anywhere()
+        foreach (var byteCheckWithoutOffset in _nonOffsetByteChecks)
+        {
+            for (var index = 1; index <= fileByteStream.Length; index++)
+            {
+                if (byteCheckWithoutOffset.Cast<byte>().SequenceEqual(fileByteStream.Skip(index).Take(byteCheckWithoutOffset.Length)))
+                    return true;
+            }
+
+            return false;
+        }
+
+        return true;
+    }
+    
+    public FileByteFilter StartsWith(byte?[] bytesToCheck)
+    {
+        _neededByteChecks.Add(new ByteCheck(0, bytesToCheck));
+        return this;
+    }
+    
+    public FileByteFilter StartsWithAnyOf(byte?[][] bytesToCheck)
+    {
+        _oneOfEachByteChecks.Add(bytesToCheck.Select(byteArray => new ByteCheck(0, byteArray)).ToArray());
+        return this;
+    }
+    
+    public FileByteFilter EndsWith(byte?[] bytesToCheck)
+    {
+        _neededByteChecks.Add(new ByteCheck(-1, bytesToCheck));
+        return this;
+    }
+    
+    public FileByteFilter EndsWithAnyOf(byte?[][] bytesToCheck)
+    {
+        _oneOfEachByteChecks.Add(bytesToCheck.Select(byteArray => new ByteCheck(-1, byteArray)).ToArray());
+        return this;
+    }
+
+    public FileByteFilter Anywhere(byte?[] bytesToCheck)
+    {
+        _nonOffsetByteChecks.Add(bytesToCheck);
+        return this;
+    }
+    
+    public FileByteFilter Anywhere(byte?[][] bytesToCheck)
+    {
+        foreach (var byteArrayToCheck in bytesToCheck)
+        {
+            Anywhere(byteArrayToCheck);
+        }
+        return this;
+    }
+
+    public FileByteFilter Specific(ByteCheck bytesToCheck)
+    {
+        _neededByteChecks.Add(bytesToCheck);
+        return this;
+    }
+    
+    public FileByteFilter SpecificAnyOf(ByteCheck[] bytesToCheck)
+    {
+        _oneOfEachByteChecks.Add(bytesToCheck.Select(byteArray => byteArray).ToArray());
+        return this;
+    }
+
+    private bool CheckBytes(ByteCheck byteToCheck, byte[] fileStreamToCheck)
+    {
+        // Check ending of file stream
+        // since in the current format we have the fileStream Length only here calculate the offset
+        if (byteToCheck.Offset == -1)
```

Comment:
Since offsets are always seen from the beginning of file, negative offsets shouldn't be possible. 
We might want to add negative offsets in the future to count offsets from the end of the file but we currently do not support those.